### PR TITLE
Refactor write_gcm create monitored resource function

### PR DESCRIPTION
**Pull request by dhrupadb**
The function currently takes an argument list which is convenient when labels are known and finite. In order to support multiple labels (possibly required for container support), moved logic to a different function which takes an array argument and had the original function call it.

### CODE REVIEW

File: src/write_gcm.c [hunk 1]
```diff
@@ -2147,57 +2147,74 @@ static monitored_resource_t *wg_monitored_resource_create(
   return NULL;
 }
 
-static monitored_resource_t *monitored_resource_create_from_fields(
-    const char *type, const char *project_id, ...) {
+
+static monitored_resource_t *monitored_resource_create_from_field_arr(
```
Comment (igorpeshansky) [inline src/write_gcm.c]: I'm not too crazy about this name. How about `monitored_resource_create_from_array`?

**Review by igorpeshansky — CHANGES_REQUESTED** [review]

File: src/write_gcm.c (hunk 1 above)
Comment (dhrupadb) [inline src/write_gcm.c]: Done

File: src/write_gcm.c [hunk 2]
```diff
@@ -2147,57 +2147,74 @@ static monitored_resource_t *wg_monitored_resource_create(
...
+    const char *type, const char *project_id, const char **keys,
```
Comment (dhrupadb) [inline src/write_gcm.c]: Done.

Comment (dhrupadb) [inline src/write_gcm.c]: Fixed.

File: src/write_gcm.c [hunk 3]
```diff
@@ -2147,57 +2147,74 @@ static monitored_resource_t *wg_monitored_resource_create(
...
-      break;
-    }
-    const char *nextValue = va_arg(ap, const char*);
-    (void)nextValue;  // unused
-    ++num_labels;
-  }
-  va_end(ap);
-
   result->num_labels = num_labels;
+
```
Comment (dhrupadb) [inline src/write_gcm.c]: Fixed.

File: src/write_gcm.c [hunk 4]
```diff
@@ -2147,57 +2147,74 @@ static monitored_resource_t *wg_monitored_resource_create(
...
+    ERROR("write_gcm: monitored_resource_create_from_field_arr: calloc failed.");
     goto error;
   }
 
-  va_start(ap, project_id);
   int i;
   for (i = 0; i < num_labels; ++i) {
-    const char *nextKey = va_arg(ap, const char*);
-    const char *nextValue = va_arg(ap, const char*);
+    const char *nextKey = keys[i];
```
Comment (dhrupadb) [inline src/write_gcm.c]: Done.

File: src/write_gcm.c [hunk 5]
```diff
@@ -2147,57 +2147,74 @@ static monitored_resource_t *wg_monitored_resource_create(
...
       goto error;
     }
   }
-  va_end(ap);
   return result;
 
  error:
   wg_monitored_resource_destroy(result);
   return NULL;
+
```
Comment (dhrupadb) [inline src/write_gcm.c]: Fixed.

File: src/write_gcm.c [hunk 6]
```diff
@@ -2147,57 +2147,74 @@ static monitored_resource_t *wg_monitored_resource_create(
...
+}
+
+static monitored_resource_t *monitored_resource_create_from_fields(
+    const char *type, const char *project_id, ...) {
+  // count keys/values
+  va_list ap;
+  va_start(ap, project_id);
+  size_t num_labels = 0;
```
Comment (dhrupadb) [inline src/write_gcm.c]: Yes, I imagined it was just convention in case it was being in that way.

File: src/write_gcm.c [hunk 7]
```diff
@@ -2147,57 +2147,74 @@ static monitored_resource_t *wg_monitored_resource_create(
...
+    ++num_labels;
+  }
+  va_end(ap);
+
+  const char *keys[num_labels];
+  const char *values[num_labels];
+  // Changes va_list into key, value arrays
+  va_start(ap, project_id);
+  int i;
+  for(i = 0; i < num_labels; i++) {
```
Comment (dhrupadb) [inline src/write_gcm.c]: Fixed.

**Review by dhrupadb — COMMENTED** [review]
PTAL

File: src/write_gcm.c [hunk 8]
```diff
@@ -2147,19 +2147,50 @@ static monitored_resource_t *wg_monitored_resource_create(
...
 
-static monitored_resource_t *monitored_resource_create_from_fields(
-    const char *type, const char *project_id, ...) {
+typedef struct {
+  const char *key;
+  const char *value;
+} label_t;
+
+static monitored_resource_t *monitored_resource_create_from_array(
+    const char *type, const char *project_id, label_t *labels,
```
Comment (igorpeshansky) [inline src/write_gcm.c]: Want to make `labels` a const pointer as well?

File: src/write_gcm.c [hunk 9]
```diff
@@ -2147,19 +2147,50 @@ static monitored_resource_t *wg_monitored_resource_create(
...
+    size_t num_labels) {
   monitored_resource_t *result = calloc(1, sizeof(*result));
   if (result == NULL) {
-    ERROR("write_gcm: monitored_resource_create_from_fields: calloc failed.");
+    ERROR("write_gcm: monitored_resource_create_from_field_arr: calloc failed.");
```
Comment (igorpeshansky) [inline src/write_gcm.c]: Update the function name? Here and below.

File: src/write_gcm.c [hunk 10]
```diff
@@ -2170,34 +2201,16 @@ static monitored_resource_t *monitored_resource_create_from_fields(
...
-
-  result->num_labels = num_labels;
-  result->keys = calloc(num_labels, sizeof(result->keys[0]));
-  result->values = calloc(num_labels, sizeof(result->values[0]));
-  if (result->keys == NULL || result->values == NULL) {
-    ERROR("write_gcm: monitored_resource_create_from_fields: calloc failed.");
-    goto error;
-  }
-
+  label_t labels[num_labels];
```
Comment (igorpeshansky) [inline src/write_gcm.c]: Let's keep the blank line between `va_end` and this declaration.

**Review by igorpeshansky — CHANGES_REQUESTED** [review]

**Comment by dhrupadb** [comment]
Done. PTAL

File: src/write_gcm.c (hunk 9 above)
Comment (igorpeshansky) [inline src/write_gcm.c]: You seem to have missed this one.

**Comment by dhrupadb** [comment]
Sorry missed that one comment. PTAL.

File: src/write_gcm.c [hunk 11]
```diff
@@ -2147,19 +2147,50 @@ static monitored_resource_t *wg_monitored_resource_create(
...
+  if (result->keys == NULL || result->values == NULL) {
+    ERROR("write_gcm: monitored_resource_create_from_array: calloc failed.");
+    goto error;
+  }
+  for (int i = 0; i < num_labels; ++i) {
+    result->keys[i] = sstrdup(labels[i].key);
+    result->values[i] = sstrdup(labels[i].value);
+    if (result->keys[i] == NULL || result->values[i] == NULL) {
+      ERROR(
+          "write_gcm: monitored_resource_create_from_array: calloc failed.");
```
Comment (igorpeshansky) [inline src/write_gcm.c]: Fits on one line now.

Comment (dhrupadb) [inline src/write_gcm.c]: Done

**Review by igorpeshansky — APPROVED** [review]

**Comment by igorpeshansky** [comment]
LGTM :shipit:

**Comment by igorpeshansky** [comment]
dhrupadb was at Google at the time.

**Pull request merged**